### PR TITLE
chore: clarify snapshot capture tests

### PR DESCRIPTION
**Pull request by pauldambra**
The snapshot capture tests only generated pre-august 2023 version snapshot events 

Well, not any more! 

See https://github.com/PostHog/product-internal/pull/525

### CODE REVIEW

File: posthog/api/test/test_capture.py [hunk 1]
```diff
@@ -134,6 +136,48 @@ def _send_session_recording_event(
...
+                "$snapshot_data": event_data,
+                "$session_id": session_id,
+                "$window_id": window_id,
+                # snapshot events have the distinct id in the properties
+                # as well as at the top-level
+                "distinct_id": distinct_id,
+            },
+            "timestamp": timestamp,
+            "distinct_id": distinct_id,
+        }
```
Comment (pauldambra) [inline posthog/api/test/test_capture.py:169]: @marandaneto 

I've defuckulated this against the JS SDK

* we send an array of PostHog events to /s/
* each of those has top level "event", "properties", "timestamp", and "distinct id" key
* the event key must be "$snapshot"
* the properties key 
    * has "$snapshot_bytes" optional size estimate
    * "$snapshot_data" array of rrweb events
    * "$session_id"
    * "$window_id"
    * "distinct_id" copied from the top level

Comment (marandaneto) [reply posthog/api/test/test_capture.py:169]: > we send an array of PostHog events to /s/

I am still confused. this is contradictory with the test:

Based on this comment, the payload should be:

`data=[the events...]`
but its actually:
`data={"data": [the events...], "api_key": "the key"}`

So the JS sends an object that contains the `api_key` and a `data` array that is an array of events, I guess? if that's the case, that's exactly what I've tried, and its failing with 400.

The example is:
```
{
  "api_key": "theApiKey",
  "data": [
    {
      "distinct_id": "e3de4e90-491f-4164-9aed-40a3d7881978",
      "event": "$snapshot",
      "properties": {
        "$screen_density": 2.75,
        "$screen_height": 2154,
        "$screen_width": 1080,
        "$app_version": "1.0",
        "$app_namespace": "com.posthog.android.sample",
        "$app_build": 1,
        "$app_name": "PostHog Android Sample",
        "$device_manufacturer": "Google",
        "$device_model": "sdk_gphone64_arm64",
        "$device_name": "emu64a",
        "$device_type": "Mobile",
        "$os_name": "Android",
        "$os_version": "14",
        "$lib": "posthog-android",
        "$lib_version": "3.0.0-beta.3",
        "$locale": "en-US",
        "$user_agent": "Dalvik/2.1.0 (Linux; U; Android 14; sdk_gphone64_arm64 Build/UPB5.230623.003)",
        "$timezone": "Europe/Vienna",
        "$network_wifi": true,
        "$network_bluetooth": false,
        "$network_cellular": false,
        "$network_carrier": "T-Mobile",
        "$snapshot_data": [
          {
            "timestamp": 1699354586963,
            "type": 0
          },
          {
            "timestamp": 1699354586963,
            "type": 1
          },
          {
            "data": {
              "href": "http://localhost",
              "width": 1080,
              "height": 2220
            },
            "timestamp": 1699354586963,
            "type": 4
          },
          {
            "data": {
              "node": {
                "id": 1,
                "type": 0,
                "childNodes": [
                  {
                    "type": 1,
                    "name": "html",
                    "id": 2,
                    "childNodes": []
                  },
                  {
                    "id": 3,
                    "type": 2,
                    "tagName": "html",
                    "childNodes": [
                      {
                        "id": 5,
                        "type": 2,
                        "tagName": "body",
                        "childNodes": [
                          {
                            "type": 2,
                            "tagName": "canvas",
                            "id": 7,
                            "attributes": {
                              "id": "canvas",
                              "width": "1080",
                              "height": "2220"
                            },
                            "childNodes": []
                          }
                        ]
                      }
                    ]
                  }
                ],
                "initialOffset": {
                  "left": 0,
                  "top": 0
                }
              }
            },
            "timestamp": 1699354586963,
            "type": 2
          },
          {
            "data": {
              "source": 9,
              "id": 7,
              "type": 0,
              "commands": [
                {
                  "property": "clearRect",
                  "args": [
                    0,
                    0,
                    1080,
                    2220
                  ]
                },
                {
                  "property": "save",
                  "args": []
                },
                {
                  "property": "translate",
                  "args": [
                    0.0,
                    2154.0
                  ]
                },
                {
                  "property": "fillStyle",
                  "args": [
                    "rgba(0, 0, 0, 255)"
                  ],
                  "setter": true
                },
                {
                  "property": "strokeStyle",
                  "args": [
                    "rgba(0, 0, 0, 255)"
                  ],
                  "setter": true
                },
                {
                  "property": "font",
                  "args": [
                    "normal 12px sans-serif"
                  ],
                  "setter": true
                },
                {
                  "property": "textAlign",
                  "args": [
                    "left"
                  ],
                  "setter": true
                },
                {
                  "property": "lineWidth",
                  "args": [
                    0.0
                  ],
                  "setter": true
                },
                {
                  "property": "globalAlpha",
                  "args": [
                    1.0
                  ],
                  "setter": true
                },
                {
                  "property": "fillRect",
                  "args": [
                    0.0,
                    0.0,
                    1080.0,
                    66.0
                  ]
                },
                {
                  "property": "restore",
                  "args": []
                },
                {
                  "property": "restore",
                  "args": []
                },
                {
                  "property": "save",
                  "args": []
                },
                {
                  "property": "translate",
                  "args": [
                    0.0,
                    0.0
                  ]
                },
                {
                  "property": "fillStyle",
                  "args": [
                    "rgba(0, 102, 139, 255)"
                  ],
                  "setter": true
                },
                {
                  "property": "strokeStyle",
                  "args": [
                    "rgba(0, 102, 139, 255)"
                  ],
                  "setter": true
                },
                {
                  "property": "font",
                  "args": [
                    "normal 12px sans-serif"
                  ],
                  "setter": true
                },
                {
                  "property": "textAlign",
                  "args": [
                    "left"
                  ],
                  "setter": true
                },
                {
                  "property": "lineWidth",
                  "args": [
                    0.0
                  ],
                  "setter": true
                },
                {
                  "property": "globalAlpha",
                  "args": [
                    1.0
                  ],
                  "setter": true
                },
                {
                  "property": "fillRect",
                  "args": [
                    0.0,
                    0.0,
                    1080.0,
                    66.0
                  ]
                },
                {
                  "property": "restore",
                  "args": []
                },
                {
                  "property": "restore",
                  "args": []
                },
                {
                  "property": "save",
                  "args": []
                },
                {
                  "property": "translate",
                  "args": [
                    0.0,
                    66.0
                  ]
                },
                {
                  "property": "restore",
                  "args": []
                },
                {
                  "property": "restore",
                  "args": []
                }
              ]
            },
            "timestamp": 1699354586963,
            "type": 3
          },
          {
            "data": {
              "source": 2,
              "type": 1,
              "id": 7,
              "x": 832.97266,
              "y": 928.9072
            },
            "timestamp": 1699354595006,
            "type": 3
          },
          {
            "data": {
              "source": 2,
              "type": 2,
              "id": 7,
              "x": 832.97266,
              "y": 928.9072
            },
            "timestamp": 1699354595068,
            "type": 3
          },
          {
            "data": {
              "source": 2,
              "type": 1,
              "id": 7,
              "x": 832.97266,
              "y": 928.9072
            },
            "timestamp": 1699354595148,
            "type": 3
          },
          {
            "data": {
              "positions": [
                {
                  "x": 826.97363,
                  "y": 928.9072,
                  "id": 7,
                  "timeOffset": 0
                }
              ]
            },
            "timestamp": 1699354595221,
            "type": 3
          },
          {
            "data": {
              "positions": [
                {
                  "x": 826.97363,
                  "y": 923.96094,
                  "id": 7,
                  "timeOffset": 0
                }
              ]
            },
            "timestamp": 1699354595235,
            "type": 3
          },
          {
            "data": {
              "source": 2,
              "type": 2,
              "id": 7,
              "x": 826.97363,
              "y": 923.96094
            },
            "timestamp": 1699354595235,
            "type": 3
          },
          {
            "data": {
              "source": 2,
              "type": 1,
              "id": 7,
              "x": 462.9746,
              "y": 798.96387
            },
            "timestamp": 1699354595328,
            "type": 3
          },
          {
            "data": {
              "positions": [
                {
                  "x": 436.36523,
                  "y": 780.94336,
                  "id": 7,
                  "timeOffset": 0
                }
              ]
            },
            "timestamp": 1699354595337,
            "type": 3
          },
          {
            "data": {
              "positions": [
                {
                  "x": 422.05176,
                  "y": 769.0293,
                  "id": 7,
                  "timeOffset": 0
                }
              ]
            },
            "timestamp": 1699354595354,
            "type": 3
          },
          {
            "data": {
              "positions": [
                {
                  "x": 413.28027,
                  "y": 751.57324,
                  "id": 7,
                  "timeOffset": 0
                }
              ]
            },
            "timestamp": 1699354595370,
            "type": 3
          },
          {
            "data": {
              "positions": [
                {
                  "x": 409.97656,
                  "y": 744.96875,
                  "id": 7,
                  "timeOffset": 0
                }
              ]
            },
            "timestamp": 1699354595383,
            "type": 3
          },
          {
            "data": {
              "source": 2,
              "type": 2,
              "id": 7,
              "x": 409.97656,
              "y": 744.96875
            },
            "timestamp": 1699354595384,
            "type": 3
          },
          {
            "data": {
              "source": 2,
              "type": 1,
              "id": 7,
              "x": 373.98535,
              "y": 643.9541
            },
            "timestamp": 1699354595472,
            "type": 3
          },
          {
            "data": {
              "positions": [
                {
                  "x": 355.93945,
                  "y": 619.9365,
                  "id": 7,
                  "timeOffset": 0
                }
              ]
            },
            "timestamp": 1699354595487,
            "type": 3
          },
          {
            "data": {
              "positions": [
                {
                  "x": 343.52344,
                  "y": 601.5117,
                  "id": 7,
                  "timeOffset": 0
                }
              ]
            },
            "timestamp": 1699354595504,
            "type": 3
          },
          {
            "data": {
              "positions": [
                {
                  "x": 325.64453,
                  "y": 565.17773,
                  "id": 7,
                  "timeOffset": 0
                }
              ]
            },
            "timestamp": 1699354595520,
            "type": 3
          },
          {
            "data": {
              "source": 2,
              "type": 2,
              "id": 7,
              "x": 325.96484,
              "y": 565.9756
            },
            "timestamp": 1699354595528,
            "type": 3
          },
          {
            "data": {
              "source": 2,
              "type": 1,
              "id": 7,
              "x": 188.9873,
              "y": 405.95215
            },
            "timestamp": 1699354595609,
            "type": 3
          },
          {
            "data": {
              "positions": [
                {
                  "x": 188.9873,
                  "y": 399.92285,
                  "id": 7,
                  "timeOffset": 0
                }
              ]
            },
            "timestamp": 1699354595621,
            "type": 3
          },
          {
            "data": {
              "positions": [
                {
                  "x": 182.98828,
                  "y": 393.96094,
                  "id": 7,
                  "timeOffset": 0
                }
              ]
            },
            "timestamp": 1699354595640,
            "type": 3
          },
          {
            "data": {
              "positions": [
                {
                  "x": 182.98828,
                  "y": 387.93066,
                  "id": 7,
                  "timeOffset": 0
                }
              ]
            },
            "timestamp": 1699354595671,
            "type": 3
          },
          {
            "data": {
              "source": 2,
              "type": 2,
              "id": 7,
              "x": 182.98828,
              "y": 387.93066
            },
            "timestamp": 1699354595677,
            "type": 3
          },
          {
            "data": {
              "source": 2,
              "type": 1,
              "id": 7,
              "x": 188.9873,
              "y": 357.98633
            },
            "timestamp": 1699354595742,
            "type": 3
          },
          {
            "data": {
              "positions": [
                {
                  "x": 194.98535,
                  "y": 357.98633,
                  "id": 7,
                  "timeOffset": 0
                }
              ]
            },
            "timestamp": 1699354595755,
            "type": 3
          },
          {
            "data": {
              "positions": [
                {
                  "x": 194.98535,
                  "y": 351.95605,
                  "id": 7,
                  "timeOffset": 0
                }
              ]
            },
            "timestamp": 1699354595772,
            "type": 3
          },
          {
            "data": {
              "positions": [
                {
                  "x": 214.51855,
                  "y": 345.1543,
                  "id": 7,
                  "timeOffset": 0
                }
              ]
            },
            "timestamp": 1699354595804,
            "type": 3
          },
          {
            "data": {
              "positions": [
                {
                  "x": 236.97559,
                  "y": 339.96484,
                  "id": 7,
                  "timeOffset": 0
                }
              ]
            },
            "timestamp": 1699354595823,
            "type": 3
          },
          {
            "data": {
              "source": 2,
              "type": 2,
              "id": 7,
              "x": 236.97559,
              "y": 339.96484
            },
            "timestamp": 1699354595824,
            "type": 3
          },
          {
            "data": {
              "source": 2,
              "type": 1,
              "id": 7,
              "x": 498.96582,
              "y": 321.94336
            },
            "timestamp": 1699354595891,
            "type": 3
          },
          {
            "data": {
              "positions": [
                {
                  "x": 536.8574,
                  "y": 321.94336,
                  "id": 7,
                  "timeOffset": 0
                }
              ]
            },
            "timestamp": 1699354595904,
            "type": 3
          },
          {
            "data": {
              "positions": [
                {
                  "x": 545.6377,
                  "y": 321.94336,
                  "id": 7,
                  "timeOffset": 0
                }
              ]
            },
            "timestamp": 1699354595920,
            "type": 3
          },
          {
            "data": {
              "positions": [
                {
                  "x": 549.47754,
                  "y": 321.94336,
                  "id": 7,
                  "timeOffset": 0
                }
              ]
            },
            "timestamp": 1699354595941,
            "type": 3
          },
          {
            "data": {
              "source": 2,
              "type": 2,
              "id": 7,
              "x": 546.9541,
              "y": 321.94336
            },
            "timestamp": 1699354595969,
            "type": 3
          },
          {
            "data": {
              "source": 2,
              "type": 1,
              "id": 7,
              "x": 624.96875,
              "y": 321.94336
            },
            "timestamp": 1699354596024,
            "type": 3
          },
          {
            "data": {
              "positions": [
                {
                  "x": 629.9785,
                  "y": 321.94336,
                  "id": 7,
                  "timeOffset": 0
                }
              ]
            },
            "timestamp": 1699354596054,
            "type": 3
          },
          {
            "data": {
              "positions": [
                {
                  "x": 629.9785,
                  "y": 315.98145,
                  "id": 7,
                  "timeOffset": 0
                }
              ]
            },
            "timestamp": 1699354596088,
            "type": 3
          },
          {
            "data": {
              "source": 2,
              "type": 2,
              "id": 7,
              "x": 629.9785,
              "y": 315.98145
            },
            "timestamp": 1699354596104,
            "type": 3
          },
          {
            "data": {
              "source": 2,
              "type": 1,
              "id": 7,
              "x": 606.97266,
              "y": 304.93848
            },
            "timestamp": 1699354596174,
            "type": 3
          },
          {
            "data": {
              "positions": [
                {
                  "x": 533.37695,
                  "y": 290.2207,
                  "id": 7,
                  "timeOffset": 0
                }
              ]
            },
            "timestamp": 1699354596188,
            "type": 3
          },
          {
            "data": {
              "positions": [
                {
                  "x": 505.77832,
                  "y": 286.98535,
                  "id": 7,
                  "timeOffset": 0
                }
              ]
            },
            "timestamp": 1699354596204,
            "type": 3
          },
          {
            "data": {
              "positions": [
                {
                  "x": 475.14453,
                  "y": 286.98535,
                  "id": 7,
                  "timeOffset": 0
                }
              ]
            },
            "timestamp": 1699354596220,
            "type": 3
          },
          {
            "data": {
              "positions": [
                {
                  "x": 466.39355,
                  "y": 280.95508,
                  "id": 7,
                  "timeOffset": 0
                }
              ]
            },
            "timestamp": 1699354596237,
            "type": 3
          },
          {
            "data": {
              "positions": [
                {
                  "x": 455.66016,
                  "y": 280.95508,
                  "id": 7,
                  "timeOffset": 0
                }
              ]
            },
            "timestamp": 1699354596254,
            "type": 3
          },
          {
            "data": {
              "source": 2,
              "type": 2,
              "id": 7,
              "x": 456.97656,
              "y": 280.95508
            },
            "timestamp": 1699354596258,
            "type": 3
          },
          {
            "data": {
              "source": 2,
              "type": 1,
              "id": 7,
              "x": 158.99414,
              "y": 95.93262
            },
            "timestamp": 1699354596683,
            "type": 3
          },
          {
            "data": {
              "source": 2,
              "type": 2,
              "id": 7,
              "x": 158.99414,
              "y": 95.93262
            },
            "timestamp": 1699354596754,
            "type": 3
          },
          {
            "data": {
              "source": 2,
              "type": 1,
              "id": 7,
              "x": 671.9678,
              "y": 613.9414
            },
            "timestamp": 1699354597017,
            "type": 3
          },
          {
            "data": {
              "source": 2,
              "type": 2,
              "id": 7,
              "x": 671.9678,
              "y": 613.9414
            },
            "timestamp": 1699354597098,
            "type": 3
          }
        ],
        "$session_id": "bceaa9ce-dc9d-4728-8a90-4a7c249604b1",
        "$window_id": "31bfffdc-79fc-4504-9ff4-0216a58bf7f6",
        "distinct_id": "e3de4e90-491f-4164-9aed-40a3d7881978"
      },
      "timestamp": "2023-11-07T10:56:46.601Z",
      "uuid": "deaa7e00-e1a4-480d-9145-fb8461678dae"
    }
  ]
}
```

and this example fails with:

```
{
  "type": "validation_error",
  "code": "invalid_payload",
  "detail": "Invalid payload: All events must have the event name field \"event\"!",
  "attr": null
}
```

Comment (pauldambra) [reply posthog/api/test/test_capture.py:169]: added that payload (reduced length of the rrweb payloads a little) and it is accepted

Comment (pauldambra) [reply posthog/api/test/test_capture.py:169]: > > we send an array of PostHog events to /s/

> I am still confused. this is contradictory with the test:

we send an array of PostHog events against a key "data" to /s/

Comment (pauldambra) [reply posthog/api/test/test_capture.py:253]: Ah ha!

We discovered that if we don't always send the data key. Particularly if you set content-type to `application/json` then the API expects that the body is the payload (not the body with a data key containing the payload)

Comment (marandaneto) [reply posthog/api/test/test_capture.py:253]: Yep, in this case, the `api_key` has to be part of the event as well otherwise you are not authorized.

Comment (pauldambra) [reply posthog/api/test/test_capture.py:253]: Well, the token can go in the query parameters too... but then you lose the signal of the token being in the event properties.

Given we don't store that data in session replay. You could omit it from the replay events and put it in the URL so we're sending less data over the wire 

So, you can send `https://blah/s/?token=the_api_token` and it'll work

Comment (marandaneto) [reply posthog/api/test/test_capture.py:253]: Is `api_token` considered sensitive? probably not I guess, the thing is, the payload is encrypted when using SSL but the URL isn't.

Anyway, fine either way, thanks for checking this out.

**Review by marandaneto — APPROVED** [review]

File: posthog/api/test/test_capture.py (hunk 1 above)
Comment (pauldambra) [reply posthog/api/test/test_capture.py:253]: It's the same token that is in plain text in your site if you're using the snippet. They're write-only and not sensitive

**Pull request merged**